### PR TITLE
#89 support closures/anonymous functions

### LINKED ISSUE #89: WP_Mock::expectHookAdded() does not support calling add_action() with a closure
**TL;DR: when I call `add_action()` with a closure callback, `expectHookAdded()` does not recognize the closure as having been called. Adding a call to `$responder->react()` inside `expectHookAdded()` fixes this.** 

Since the real `add_action()` is designed to take any callable, I would expect to be able to call `expectHookAdded` with any valid callable, including a closure.

For example, I am trying to unit test a trait that encapsulates adding an admin filter on the Posts listing screen. Specifically, I'm testing the trait's `addAdminFilter()` method. I would expect to be able to do something like this:

```php
<?php

require_once 'vendor/autoload.php';

use PHPUnit\Framework\TestCase;

trait HasCustomAdminFilters {
  public function addAdminFilter( $name, $options, $postType, callable $queryModifier ) {
    // allow user to define how they want the admin query modified
    add_action('pre_get_posts', function(\WP_Query $query) use($name, $postType, $queryModifier) {
      // call the user-defined query modifying code
      $queryModifier($query);
    });
  }
}

class HasCustomAdminFiltersTest extends TestCase {
  public function setUp() {
    WP_Mock::setUp();
  }

  public function tearDown() {
    WP_Mock::tearDown();
  }

  public function testAddAdminFilter() {
    $callback = function() {};

    WP_Mock::expectActionAdded('pre_get_posts', WP_Mock\Functions::type('callable'));

    // register the action handlers
    $this->getObjectForTrait('HasCustomAdminFilters')->addAdminFilter(
      'custom_filter',
      ['dropdown' => 'options'],
      'my_post_type',
      $callback
    );
  }
}
```

This throws:

> Mockery\Exception\InvalidCountException: Method intercepted() from Mockery_0__intercept should be called at least 1 times but called 0 times.

Digging into the `WP_Mock` class, I see that the `expectHookAdded()` code creates an internal mock object and tells it to be `intercepted`; then it "performs" the added hook.

This works for the basic use-case of declaring a global function, e.g. `foo()`, and then calling `add_action('pre_get_posts', 'foo')`. But in my use-case, I want to encapsulate my callback as a closure, not expose it as a global function. Adding a call to `$responder->react()` at the end of `expectHookAdded()` fixes this issue.

I'm not sure if such a change would have other implications. Is it an appropriate solution? Should `expectAction()` also get the same treatment?

### PR DESCRIPTION
**Pull request by acobster**
This fixes https://github.com/10up/wp_mock/issues/89 and supsedes PRs #130 and https://github.com/10up/wp_mock/pull/119.

The root of the issue turned out to be that using WP_Mock\Functions::type('callable') was causing a mismatch between the safe_offset values returned for the Mockery matcher and the actual closure being passed to add_action. The matcher was getting spl_object_hash'd, while the Closure was getting `"__CLOSURE__"`. Now, safe_offset() treats Mockery matchers with type closure as `"__CLOSURE__"` too.

### CODE REVIEW

**Comment by acobster** [comment]
OK, I think we should be all set here. Let me know if there's anything else I can do.

File: features/bootstrap/HooksContext.php [hunk 1]
```diff
@@ -7,229 +7,261 @@
...
-			'priority'  => 10,
-			'arguments' => 1,
-		);
-		foreach ( $filters as &$filter ) {
-			$filter += $defaults;
-		}
-		unset( $filter );
-
-		return $filters;
-	}
```
Comment (unfulvio-godaddy) [inline features/bootstrap/HooksContext.php:233]: looks like this portion of the file was accidentally formatted perhaps by your editor- we plan to introduce some codestyle / phpcs automation to avoid style inconsistencies in the future - would you mind restoring the lines that have no code/logic change please?

File: php/WP_Mock/Hook.php [hunk 2]
```diff
@@ -25,6 +25,8 @@ protected function safe_offset( $value ) {
 			return 'null';
 		} elseif ( is_scalar( $value ) ) {
 			return $value;
+		} elseif ( ''.$value === '<Closure>' ) {
+			return '__CLOSURE__';
```
Comment (unfulvio-godaddy) [inline php/WP_Mock/Hook.php:29]: hey @acobster may I know why the concatenation with `''.$value` here? shouldn't you check `is_string( $value ) && $value === '<Closure>'` instead? is this expected to have a toString method? or isn't this expected to be `$value instanceof \Closure`? Just double checking, I've read your previous explanation in your PR description.

On the other hand, if it's a string instead, wouldn't it return true for the previous condition `is_scalar()`? 

sorry if I missed anything here

**Review by unfulvio-godaddy — CHANGES_REQUESTED** [review]

File: README.md [hunk 3]
```diff
@@ -332,6 +332,20 @@ public function test_special_function() {
...
 It's important to note that the `$priority` and `$parameter_count` arguments (parameters 3 and 4 for both `add_action()` and `add_filter()`) are significant. If `special_function()` were to call `add_action( 'save_post', 'special_save_post', 99, 3 )` instead of the expected `add_action( 'save_post', 'special_save_post', 10, 2 )`, our test would fail.
 
+#### Asserting that Closure hooks have been added
+
+Sometimes it's handy to add a [Closure](https://secure.php.net/manual/en/class.closure.php) as a WordPress hook instead of defining a function in the global namespace. To assert that such a hook has been added, you can use the special `closure` type:
+
+```php
+public function test_anonymous_function_hook() {
+	\WP_Mock::expectActionAdded( 'save_post', \WP_Mock\Functions::type( 'closure' ) );
+	\WP_Mock::expectFilterAdded( 'the_content', \WP_Mock\Functions::type( 'closure' ) );
```
Comment (wvega-godaddy) [inline README.md:342]: Should these use `type( 'callable' )` instead?

```suggestion
	\WP_Mock::expectActionAdded( 'save_post', \WP_Mock\Functions::type( 'callable' ) );
	\WP_Mock::expectFilterAdded( 'the_content', \WP_Mock\Functions::type( 'callable' ) );
```

Comment (wvega-godaddy) [reply README.md:342]: Or are we trying to match instances of the `Closure` class? In that case I think using `Clousure::class` is more accurate:

```suggestion
	\WP_Mock::expectActionAdded( 'save_post', \WP_Mock\Functions::type( Closure::class ) );
	\WP_Mock::expectFilterAdded( 'the_content', \WP_Mock\Functions::type( Closure::class ) );
```

`\WP_Mock\Functions::type( 'closure' )` works because `is_class( 'closure' )` returns `true` as class names are not case sensitive, but it took me a lot of digging to understand that.

Initially I thought there was some special handling for the `closure` string, but we are [passing that string to `Mockery::type();`](https://github.com/10up/wp_mock/blob/d0584171d602e60d1d58e2e0dccd35dcce529f18/php/WP_Mock/Functions.php#L99-L112) and [it is used as a class name](https://github.com/mockery/mockery/blob/3c3ca93df1752eb33ea9cfa9481c974e650b604c/library/Mockery/Matcher/Type.php#L40-L43).

If we are targeting the class name, I believe using `Closure::class` better reflects what's happening under the hood and makes the example more readable.

File: php/WP_Mock/Hook.php (hunk 2 above)
Comment (wvega-godaddy) [reply php/WP_Mock/Hook.php:29]: The code in `trunk` has [a new case to check for instances of `AnyInstance`](https://github.com/10up/wp_mock/blob/d0584171d602e60d1d58e2e0dccd35dcce529f18/php/WP_Mock/Hook.php#L30-L31) in this method.

Maybe this PR can check for [instances of `Mockery\Matcher\Type`](https://github.com/mockery/mockery/blob/3c3ca93df1752eb33ea9cfa9481c974e650b604c/library/Mockery/Matcher/Type.php#L23), which is what we will receive when `WP_Mock\Functions::type('callable')` is passed as the expected callback?

**Comment by acobster** [comment]
@unfulvio-godaddy @wvega-godaddy I don't work on WP stuff with any regularity anymore, and since this isn't a super quick fix I don't anticipate having time to look into these in the foreseeable future. The suggested changes sound reasonable though. Thanks for your work!

File: php/WP_Mock/Hook.php (hunk 2 above)
Comment (unfulvio-godaddy) [reply php/WP_Mock/Hook.php:29]: @wvega-godaddy I've updated the code with your feedback in a separate branch `feature/closures-bis` mind checking if that's what you had in mind?

File: README.md (hunk 3 above)
Comment (unfulvio-godaddy) [reply README.md:342]: @wvega-godaddy I have applied some of your suggestions in `feature/closures-bis`

**Comment by unfulvio-godaddy** [comment]
Thank you for letting us know @acobster, no worries, we'll try to apply the change this PR originally was meant for.

@wvega-godaddy I have pushed a separate branch `feature/closures-bis` where I try to apply your feedback, after I got a bit more understanding of the components in action here. 

I think the recent addition of the `AnyInstance` matcher has caused some Behat tests originally added by this PR to fail. 

If we want to support just the mapping of `<Closure>` to `__CLOSURE__` for edge cases, I believe we still can.

File: README.md (hunk 3 above)
Comment (unfulvio-godaddy) [reply README.md:342]: @wvega-godaddy see https://github.com/10up/wp_mock/pull/183

**Comment by unfulvio-godaddy** [comment]
see https://github.com/10up/wp_mock/pull/183

**Comment by BrianHenryIE** [comment]
> I think the recent addition of the AnyInstance matcher has caused some Behat tests originally added by this PR to fail.

My apologies, even the PhpUnit test suite wasn't passing for me at the time.

I think the API of AnyInstance (or anything really, since it's pre 1.0) is open to redesign. I definitely noticed the Closure issue as I wrote/waited for the AnyInstance PR. 

I think maybe ideally, both would be abstracted behind `\WP_Mock\Functions::type( 'int' )`.

**Comment by unfulvio-godaddy** [comment]
Superesed by https://github.com/10up/wp_mock/pull/183

I have imported changes proposed by this PR into https://github.com/10up/wp_mock/pull/183 and updated unit tests

Thanks again @acobster for contributing!

I am going to close this but please feel free to comment for feedback over this issue or future issues. Happy new year!
